### PR TITLE
Microcosms microcosms

### PR DESCRIPTION
**Pull request by openbrian**
Since the first PR (#2390) was too large, this is "take two".

This is the first of four PRs.  Here we focus on just the microcosm resource.

To see the partitioning of the entire feature, see https://wiki.openstreetmap.org/wiki/Microcosms.  Feedback welcome on both this PR and the wiki page.

Mostly standard Rails stuff, but here are some highlights:
* micrososms have a friendly url (aka slug)  /microcosm/greater_dc_area based on the name of the microcosm
* microcosms have 1 organizer at this time
* access control for updating your microcosm is controlled by cancancan
* area of influence is established by zooming in on a map
* some code is already factored out because later stages will also use the same functions
* spammy microcosms may be reported

To see a demo of these features (and more), see https://microcosms.apis.dev.openstreetmap.org/ (currently down) and https://mappingdc.org/microcosms (slow server).

While I prefer the name "microcosm", I'm not bent on it.  If we all decide on a different name, I'm ready to do a global search and replace.

### CODE REVIEW

**Comment by pnorman** [comment]
I don't think we should be using the name Microcosms. It doesn't fit with the other naming of features on the site, and raises questions of how to translate it.

**Comment by openbrian** [comment]
> I don't think we should be using the name Microcosms. It doesn't fit with the other naming of features on the site, and raises questions of how to translate it.

@pnorman Do you have a suggestion?

**Comment by mmd-osm** [comment]
> https://mappingdc.org/microcosms (slow server)

Server says the service is currently unavailable. Looking for some alternatives to try this out…

**Comment by tomhughes** [comment]
I've fixed https://microcosms.apis.dev.openstreetmap.org/.

**Comment by pnorman** [comment]
> @pnorman Do you have a suggestion?

At SOTM US in Minneapolis I thought we agreed on "Groups"?

**Comment by mmd-osm** [comment]
> At SOTM US in Minneapolis I thought we agreed on "Groups"?

Maybe worth keeping in mind: next door in #3301, another top level navigation link "Communities" is about to be added.

"Communities" and "Groups" right next to each other may be somewhat confusing, in particular, if a translation ends up using the same word for both.

**Comment by openbrian** [comment]
Not an issue.  Adam and I have been chatting.  Literally a video chat and screen share.  We're collaborating on this.

**Review by gravitystorm — CHANGES_REQUESTED** [review]
I've had a first pass at reviewing this PR, but even at 52 files it's still a lot of stuff in one go.

There's a few things that stand out as also needing more thought on my side, beyond what I've put into the comments:

* Is there any way to simplify the PR further. For example, perhaps we can cut out the issue reporting for now, and add it as a follow up PR
* Given that we might want to decouple reviewing PRs incrementally from launching the whole feature, do we need to put this behind a feature flag?
* I haven't reviewed the friendly_id stuff, it seems to add a lot of complexity and doesn't follow the "id-friendly-text" format that I would expect (e.g. `/foo/123-name-here`) for stable links.
* We need to think through the "one organizer for now" bit of this, and make sure we're not storing up trouble for later :smile: Either in terms of migrations or model relationships etc.

Thanks for all your work so far on this.

**Comment by openbrian** [comment]
Andy, thanks for your detailed review!  This is a response to your opening issues.

* Reducing the PR even further - Please understand I have already cut out membership, events, and RSVP.  Seems like you got through half of them today.
* I agree about the feature flag.  Please highlight a design pattern for that, or suggest how to do that.  There is the option to branch from this branch, and merge everything when it's all ready.
* friendly_id appears to be working correctly.  Here's a sample.  https://microcosms.apis.dev.openstreetmap.org/microcosms/openstreetmap-salt-lake-city
* The only reason we have one organizer at this time is to support access control for edit/update.  When we get proper members, they will have roles, organizer/member/whatever.  So there can be multiple organizers.  For phase 2 we create a new model/table and migrate microcosm.organizer into microcosm_member.user_id.

I'll now address all your other concerns above :)

File: app/assets/stylesheets/microcosms.scss
Comment (openbrian) [reply app/assets/stylesheets/microcosms.scss:7]: My code predates bootstrap.  I'll review all the CSS rules.

**Comment by tomhughes** [comment]
Andy's point about the friendly ids is that the normal pattern is to include the numeric id as well, so something like `.../9999/friendly-slug` so that the routing can just ignore the text part and use the number - that also means people can change the textual name as much as they like without breaking URLs.

As to the question of organisers - what is the point of creating the table one way only to change it five minutes later? We don't need to preserve the entire development history for an unmerged feature - until a feature is merged it's fine to go back and change things. A first PR that added the complete data model and then PRs that incrementally made use of that would be a great approach for example.

File: app/assets/stylesheets/microcosms.scss
Comment (openbrian) [reply app/assets/stylesheets/microcosms.scss:10]: These labels are not form labels.  The form labels are built-in to the fields and rendered via bootstrap_form_for, so they'll get default styling.

File: app/views/microcosms/index.html.erb
Comment (openbrian) [inline app/views/microcosms/index.html.erb:10]: I just spotted the All Traces / My Traces design pattern.  I'll apply that here.

File: app/assets/stylesheets/microcosms.scss
Comment (openbrian) [reply app/assets/stylesheets/microcosms.scss:12]: Bootstrap list-group looks bad, due to the opinionated styles.  The breadcrumb component works well.

File: app/assets/stylesheets/microcosms.scss
Comment (openbrian) [reply app/assets/stylesheets/microcosms.scss:20]: done

File: app/assets/stylesheets/microcosms.scss
Comment (openbrian) [reply app/assets/stylesheets/microcosms.scss:12]: done

File: config/routes.rb
Comment (openbrian) [reply config/routes.rb]: done

File: app/models/user.rb
Comment (openbrian) [reply app/models/user.rb:86]: Later when we have members, the user.microcosms will be the list of microcosms that a user belongs to.  :microcosms_i_organize is different.  It's also temporary.

I waffled on this myself.  I considered :microcosms_organized.

File: config/routes.rb
Comment (openbrian) [reply config/routes.rb:327]: Right, in this stage, there are no members yet.  So the relationship between users and microcosms is that, the user is the owner of the microcosm.  Once we have proper membership, we'll have multiple organizers of a microcosm and general membership in a microcosm.  I feel /user/Joe/microcosms would list both groups organized and groups Joe belongs to.  That's the design pattern I see on other websites out there.  I'm not opposed to creating a separate route for organizers and members.

Yeah, I feel this of_user action is in the wrong place. The microcosm links was a clear cut case.  But the users stuff -- as with diary entries, traces, and notes -- is different due to the many-to-many relationships.  The index action for notes, traces, and diary entries takes an optional :display_user. I'll get microcosms working that way.  But in routes.rb, they all do it like this.

```
  get "/user/:display_name/notes" => "notes#index", :as => :user_notes
  get "/user/:display_name/traces" => "traces#index"
  get "/user/:display_name/diary" => "diary_entries#index"
```

Seems like all of these could be nested resources.

File: app/controllers/microcosms_controller.rb
Comment (openbrian) [reply app/controllers/microcosms_controller.rb]: done

File: app/controllers/microcosms_controller.rb
Comment (openbrian) [reply app/controllers/microcosms_controller.rb]: done

File: app/controllers/microcosms_controller.rb
Comment (openbrian) [reply app/controllers/microcosms_controller.rb]: done

File: app/controllers/microcosms_controller.rb
Comment (openbrian) [reply app/controllers/microcosms_controller.rb]: done, as it's not specific to microcosms, I moved it to lib/osm.rb so it can be shared by any model with a longitude.

File: app/controllers/microcosms_controller.rb
Comment (openbrian) [reply app/controllers/microcosms_controller.rb:64]: done

File: app/views/microcosm_links/_form.html.erb
Comment (openbrian) [reply app/views/microcosm_links/_form.html.erb:2]: I removed the extra text. done

File: app/views/microcosms/_form.html.erb
Comment (openbrian) [reply app/views/microcosms/_form.html.erb:5]: done

File: app/views/microcosms/_form.html.erb
Comment (openbrian) [reply app/views/microcosms/_form.html.erb:9]: done

File: app/models/microcosm_link.rb
Comment (openbrian) [reply app/models/microcosm_link.rb:20]: Yes, It was "site" because older versions of this had links to twitter, facebook, etc. "sites".  I don't like "name".  How about "link" or "text"?  Looks like the official name is hyperlink.

File: app/views/microcosms/_empty_list.html.erb
Comment (openbrian) [reply app/views/microcosms/_empty_list.html.erb:1]: While I was going to use the `render @thing || "empty_list"` approach, I went a different route.  This file is not needed.

File: test/models/microcosm_test.rb
Comment (openbrian) [reply test/models/microcosm_test.rb]: done

File: test/test_helper.rb
Comment (openbrian) [reply test/test_helper.rb]: done

File: db/migrate/20220820220849_add_slug_to_microcosms.rb
Comment (openbrian) [reply db/migrate/20220820220849_add_slug_to_microcosms.rb]: Consolidated migrations.  Done.

File: db/migrate/20220820221326_remove_key_from_microcosms.rb
Comment (openbrian) [reply db/migrate/20220820221326_remove_key_from_microcosms.rb]: done

File: app/models/microcosm.rb
Comment (openbrian) [reply app/models/microcosm.rb:28]: Sir gravitystorm, While I disagree in this case (microcosm.organizer tells me something and microcosm.user_id tells me nothing), it doesn't matter.  It will be OBE (overcome by events, not order of the british empire) when membership arrives.

File: app/models/user.rb
Comment (openbrian) [reply app/models/user.rb]: I've removed "I".

**Comment by openbrian** [comment]
> Andy's point about the friendly ids is that the normal pattern is to include the numeric id as well, so something like `.../9999/friendly-slug` so that the routing can just ignore the text part and use the number - that also means people can change the textual name as much as they like without breaking URLs.

I've haven't seen that pattern (including ids).  WSJ, NYT, The Guardian, CNN, AP don't include ids.  When Using FriendlyId, the resource maintains it's original id attribute and is still accessible at the `GET /resource/:id route`.  When using the slug, the developers use code like this to lookup objects.

```ruby
Article.friendly.find(params[:id])
```

> As to the question of organisers - what is the point of creating the table one way only to change it five minutes later?

Without an owner for the microcosm, there's no access control for edit and update.  I wanted to include those CRUD features in this PR.  Stage 2 adds membership which supports multiple organizers.

Since the PR was broken up into 4 stages, I'm not sure what the plan is.  Here's my suggestion.  I will branch (stage 2) from this branch can call it "microcosm - membership".  After approval, we merge membership into this branch.

After stage 2, we can merge into master or wait for stage 3 (events and event organizers) and stage 4 (event attendance and rsvp).  I leave that up to you.  Membership alone does add some value, like seeing changesets of the area (which prompt for review), seeing traces, notes, and diary entries of the area.  Members can see other members and add them as friends.  And maybe even a message blast from the organizer to members.

**Comment by tomhughes** [comment]
Well counter examples of site that do use both include:

* https://gis.stackexchange.com/questions/440078/nonsense-here
* https://community.openstreetmap.org/t/nonsense-here/2987

both of which just redirect to the correct slug because they actually use the number to do the lookup.

**Comment by openbrian** [comment]
I chatted with Adam @atomoil and we're in agreement about renaming Microcosms to Communities.  The "community" in his PR will likely be renamed to CommunityIndex.

**Comment by tomhughes** [comment]
Not to forget the 1000lb gorilla that is amazon of course, where you can strip the slug from the URL because the ASIN is the real so that these are equivalent:

* https://www.amazon.co.uk/OpenStreetMap-Jonathan-Bennett-ebook/dp/B0057WR2NE
* https://www.amazon.co.uk/dp/B0057WR2NE

**Comment by openbrian** [comment]
Looks like when I rename a branch, github makes me submit a new PR.  See https://github.com/openstreetmap/openstreetmap-website/pull/3717 for "communities".

**Comment by gravitystorm** [comment]
Superseded by #3717

**Pull request closed without merging**
